### ISSUE TITLE
CSharpier causing Visual Studio responsiveness issues

### ISSUE THREAD

**Issue by davkean** [belav/csharpier#611]
Hey, I'm from the Visual Studio team and our telemetry has captured a problematic pattern inside CSharpier that causes it to consume large amounts of thread pool threads in the wild. This in turn increases the amount of time it takes for built-in Visual Studio features to respond to user actions, such as IntelliSense completion, Find, refactorings and other code actions. 

The issue was seen with all version of Visual Studio 2022 RTM.

In one dump that we captured, CSharpier was consuming over 330 thread pool threads all with the following stack. At the time we captured this dump, our telemetry told us that it taking 40 seconds from the time a work item got queued to the thread pool, to the actual time it was actually run on a real thread. Visual Studio would have been slowed to a crawl at the time this was captured and rather unpleasant to use.

```
ntdll!ZwReadFile | 100%
KERNELBASE!ReadFile | 100%
mscorlib!System.IO.FileStream.ReadFileNative(Microsoft.Win32.SafeHandles.SafeFileHandle, Byte[], Int32, Int32, System.Threading.NativeOverlapped*, Int32 ByRef) | 100%
mscorlib!System.IO.FileStream.ReadCore(Byte[], Int32, Int32) | 100%
mscorlib!System.IO.FileStream.Read(Byte[], Int32, Int32) | 100%
mscorlib!System.IO.StreamReader.ReadBuffer() | 100%
mscorlib!System.IO.StreamReader.Read() | 100%
CSharpier_VisualStudio!Unknown | 100%
mscorlib!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean) | 100%
mscorlib!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean) | 100%
mscorlib!System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem() | 100%
mscorlib!System.Threading.ThreadPoolWorkQueue.Dispatch() | 100%
clr!CallDescrWorkerInternal | 100%
clr!CallDescrWorkerWithHandler | 100%
clr!MethodDescCallSite::CallTargetWorker | 100%
clr!MethodDescCallSite::Call_RetBool | 100%
clr!QueueUserWorkItemManagedCallback | 100%
clr!ManagedThreadBase_DispatchInner | 100%
clr!ManagedThreadBase_DispatchMiddle | 100%
clr!ManagedThreadBase_DispatchOuter | 100%
clr!ManagedThreadBase_FullTransitionWithAD | 100%
clr!ManagedThreadBase::ThreadPool | 100%
clr!ManagedPerAppDomainTPCount::DispatchWorkItem | 100%
clr!ThreadpoolMgr::ExecuteWorkRequest | 100%
clr!ThreadpoolMgr::WorkerThreadStart | 100%
clr!Thread::intermediateThreadProc | 100%
kernel32!BaseThreadInitThunk | 100%
ntdll!RtlUserThreadStart
```
Because our analysis doesn't capture heap dumps, we're unable to resolve the frame referring to CSharpier, however, in the code base there is only a single consumer of StreamReader.Read:

https://github.com/belav/csharpier/blob/cf43c06dc64a1cf0ffb9a76462287d876b844893/Src/CSharpier.VisualStudio/CSharpier.VisualStudioShared/CSharpierProcessPipeMultipleFiles.cs#L104-L113

I've walked through this code and I don't see many paths that would result in >300 threads as it appears that all entry points wait, on the UI thread, for a single document to be formatted before giving back control. It doesn't appear as though these can overlap. 

However, digging deeper, I see a logic bug; both `ReadError` and `ReadOutput` share the same `AutoResetEvent`. The UI thread only ends blocking on whomever reads their output first, leaving the other thread orphaned until it has finished reading its output. I could imagine that with a sufficient amount of documents open, saving all documents (maybe many times in succession or via a solution build) could quickly flood the thread pool. To fix this, it seems like this path should coordinate with two different events and wait for them both.

**Comment by belav** [comment]
I tested this out more and version that was aborting threads would cause exceptions to be thrown because it was interrupting the `StreamReader.Read`.
The code to read from StandardError doesn't actually report any errors from the CLI due to another problem, so I am just going to remove the threads.
I released the version without threads as 1.2.1

**Comment by davkean** [comment]
Yes, Thread.Abort is not supported anymore, it can malform both static and instance state. Thanks for addressing this quickly.
